lib/core/src/Cardano/Pool/Metrics.hs

```diff
@@ -79,6 +90,74 @@ data Block = Block
     , producer :: PoolId
     } deriving (Eq, Show, Generic)
 
+--------------------------------------------------------------------------------
+-- StakePoolLayer
+--------------------------------------------------------------------------------
+
+-- | @StakePoolLayer@ is a thin layer ontop of the DB. It is /one/ value that
+-- can easily be passed to the API-server, where it can be used in a simple way.
+newtype StakePoolLayer m = StakePoolLayer
+      { listStakePools
+          :: ExceptT ErrListStakePools m
+             [(PoolId, (Quantity "lovelace" Word64, Quantity "block" Natural))]
+      }
+
+data ErrListStakePools
+     = ErrMetricsIsUnsynced (Quantity "percent" Percentage)
+     | ErrListStakePoolsMetricsInconsistency ErrMetricsInconsistency
+
+newStakePoolLayer
+     :: DBLayer IO
+     -> NetworkLayer IO t block
+     -> Trace IO Text
+     -> IO (StakePoolLayer IO)
+newStakePoolLayer db nl tr = do
+     return $ StakePoolLayer
+        { listStakePools = do
+            nodeTip <- withExceptT undefined $ networkTip nl
+            let nodeEpoch = nodeTip ^. #slotId . #epochNumber
+
+            distr <- liftIO $ Map.fromList <$> readStakeDistribution db nodeEpoch
+            prod <- liftIO $ count <$> readPoolProduction db nodeEpoch
+
+            when (Map.null distr || Map.null prod) $ do
+                mDbTip <- liftIO poolProductionTip
+                case mDbTip of
+                    Nothing -> throwE $ ErrMetricsIsUnsynced (Quantity minBound)
+                    Just dbTip -> do
+                        liftIO $ logInfo tr $ mconcat
+                            [ "The node tip is:\n"
+                            , pretty nodeTip
+                            , ",\nbut the last pool production stored in the db"
+                            , " is from:\n"
+                            , pretty dbTip
+                            ]
+                        let p = progress dbTip nodeTip
+                        throwE $ ErrMetricsIsUnsynced (Quantity p)
```

Comment:
I'd suggest to move the whole `when` body here to a separate function to make the top-level one more readable

```hs
when (Map.null dist || Map.null prod) $
    computeProgress >>= throwE . ErrMetricsIsUnsynced
```